pkg/controller/aws/network/internetgateway.go

```diff
@@ -0,0 +1,234 @@
+		resource.ManagedKind(v1alpha1.InternetGatewayGroupVersionKind),
+		resource.WithExternalConnecter(&igConnector{client: mgr.GetClient()}),
+		resource.WithManagedConnectionPublishers())
+
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1alpha1.InternetGatewayKindAPIVersion, v1alpha1.Group))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.InternetGateway{}).
+		Complete(r)
+}
+
+type igConnector struct {
+	client      client.Client
+	newClientFn func(*aws.Config) (ec2.InternetGatewayClient, error)
+}
+
+func (conn *igConnector) Connect(ctx context.Context, mgd resource.Managed) (resource.ExternalClient, error) {
+	cr, ok := mgd.(*v1alpha1.InternetGateway)
+	if !ok {
+		return nil, errors.New(errObjectNotInternetGateway)
+	}
+
+	newClientFn := ec2.NewInternetGatewayClient
+	if conn.newClientFn != nil {
+		newClientFn = conn.newClientFn
+	}
+
+	awsconfig, err := utils.RetrieveAwsConfigFromProvider(ctx, conn.client, cr.Spec.ProviderReference)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := newClientFn(awsconfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create a new InternetGatewayClient")
+	}
+
+	return &igExternal{c}, nil
+}
+
+type igExternal struct {
+	client ec2.InternetGatewayClient
+}
+
+func (e *igExternal) Observe(ctx context.Context, mgd resource.Managed) (resource.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.InternetGateway)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errObjectNotInternetGateway)
+	}
+
+	// To find out whether an InternetGateway exist:
+	// - the object's ExternalState should have internetGatewayID populated
+	// - an InternetGateway with the given internetGatewayID should exist
+	if cr.Status.InternetGatewayID == "" {
+		return resource.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	req := e.client.DescribeInternetGatewaysRequest(&awsec2.DescribeInternetGatewaysInput{
+		InternetGatewayIds: []string{cr.Status.InternetGatewayID},
+	})
+	req.SetContext(ctx)
+
+	response, err := req.Send()
+
+	if err != nil {
+		if ec2.IsInternetGatewayNotFoundErr(err) {
+			return resource.ExternalObservation{
+				ResourceExists: false,
+			}, nil
+		}
+
+		return resource.ExternalObservation{}, errors.Wrapf(err, "Error describing InternetGateway with id: %v", cr.Status.InternetGatewayID)
+	}
+
+	// in a successful response, there should be one and only one object
+	if len(response.InternetGateways) != 1 {
+		return resource.ExternalObservation{}, errors.Errorf("Retrieved multiple InternetGateways for the given internetGatewaysId: %v", cr.Status.InternetGatewayID)
+	}
+
+	observed := response.InternetGateways[0]
+
+	// if non of the attachments are currently in progress, then the IG is available
+	isAvailable := true
+	for _, a := range observed.Attachments {
+		if a.State == awsec2.AttachmentStatusAttaching || a.State == awsec2.AttachmentStatusDetaching {
+			isAvailable = false
+			break
+		}
+	}
```

Comment:
I wonder if we should set a bespoke Attaching or Detaching condition on the resource to reflect these states. e.g.:

```go
func Attaching() Condition {
	return Condition{
		Type:               TypeReady,
		Status:             corev1.ConditionFalse,
		LastTransitionTime: metav1.Now(),
		Reason:             ReasonAttaching,
	}
}
```

Reply:
I am not sure how generic such condition is going to be. I will be wary of defining too many conditions in the GMRR where each only is applicable to a small set of resource types.